rust-symcrypt/src/hash.rs

```diff
@@ -381,13 +391,26 @@ pub struct Sha256State(Pin<Box<symcrypt_sys::SYMCRYPT_SHA256_STATE>>);
 
 impl Sha256State {
     pub fn new() -> Self {
-        let mut instance = Sha256State(Box::pin(symcrypt_sys::SYMCRYPT_SHA256_STATE::default()));
+        let mut instance = Sha256State(Box::pin(Sha256InnerState{inner: symcrypt_sys::SYMCRYPT_SHA256_STATE::default(), _pinned: PhantomPinned}));
         unsafe {
             // SAFETY: FFI calls
-            symcrypt_sys::SymCryptSha256Init(&mut *instance.0);
+            symcrypt_sys::SymCryptSha256Init(instance.get_inner_mut());
         }
         instance
     }
+
+    // Safe method to access the inner state mutably
+    pub(crate) fn get_inner_mut(&mut self) -> *mut symcrypt_sys::SYMCRYPT_SHA256_STATE {
+        unsafe { 
+            // SAFETY: getting a mutable raw pointer to the inner state
+            &mut self.0.as_mut().get_unchecked_mut().inner as *mut _
+        }
+    }
```

Comment:
This function makes me a little nervous, but I think it's technically correct.

What saves you is that you are returning a `*mut` instead of an `&mut`. If this returned `&mut`, you'd be able to use it to get around the restrictions `Pin` tries to impose. But, since you return `*mut`, dereferencing that pointer is unsafe.

I would propose a couple of things to make this more robust.

First, we need to document the fact that the `*mut` that's returned is also pinned, even though that's not carried in the type. This means users must be careful not to move out of that pointer. As I mentioned before, doing so would require an `unsafe` block, but it's still good practice to document what requirements we're placing on unsafe code. I'd suggest something like:

```suggestion
    /// Get a mutable pointer to the inner SymCrypt state
    ///
    /// This is primarily meant to be used while making calls to the underlying SymCrypt APIs.
    ///
    /// ## Safety:
    /// 
    /// This function returns pointer to pinned data, which means callers must not use the pointer to move the data out of its location.
    fn get_inner_mut(&mut self) -> *mut symcrypt_sys::SYMCRYPT_SHA256_STATE {
        unsafe { 
            // SAFETY: getting a mutable raw pointer to the inner state
            &mut self.0.as_mut().get_unchecked_mut().inner as *mut _
        }
    }
```

Second, is it possible to remove the `pub(crate)` part of this? Keeping this function as tightly scoped as possible can help limit the opportunities to misuse it. I made that change in my diff above because it looks like right now this function is only used in this file.